pkg/konnector/controllers/cluster/serviceexport/cluster-scoped/utils.go

```diff
@@ -0,0 +1,123 @@
+/*
+Copyright 2023 The Kube Bind Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package clusterscoped
+
+import (
+	"errors"
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+func Prepend(name, clusterNs string) string {
+	return clusterNs + "-" + name
+}
+
+func Behead(name, clusterNs string) string {
+	return strings.TrimPrefix(name, clusterNs+"-")
+}
+
+// SwitchToUpstreamName switches the name of a downstream
+// cluster-scoped object by prepending the cluster namespace name
+func SwitchToUpstreamName(obj *unstructured.Unstructured, clusterNs string) {
+	downstreamName := obj.GetName()
+	upstreamName := Prepend(downstreamName, clusterNs)
+	obj.SetName(upstreamName)
+}
+
+// SwitchToDownstreamName switches the name of a upstream
+// cluster-scoped object by removing the cluster namespace name as the prefix
+func SwitchToDownstreamName(obj *unstructured.Unstructured, clusterNs string) {
+	upstreamName := obj.GetName()
+	downstreamName := Behead(upstreamName, clusterNs)
+	obj.SetName(downstreamName)
+}
+
+// InjectClusterNs injects the corresponding cluster namespace, by design as the
+// single owner reference, into a cluster-scoped object.
+func InjectClusterNs(obj *unstructured.Unstructured, clusterNs, clusterNsUID string) error {
+	ors := obj.GetOwnerReferences()
+	if len(ors) != 0 {
+		return errors.New("object already owned by something else")
+	}
+	ors = append(ors, metav1.OwnerReference{
+		APIVersion: "v1",
+		Kind:       "Namespace",
+		Name:       clusterNs,
+		UID:        types.UID(clusterNsUID),
+	})
+	obj.SetOwnerReferences(ors)
+	return nil
+}
+
+// ExtractClusterNs extracts the corresponding cluster namespace name
+// from a cluster-scoped object.
+func ExtractClusterNs(obj *unstructured.Unstructured) (string, error) {
+	ors := obj.GetOwnerReferences()
+	if len(ors) == 0 {
+		return "", errors.New("owner references not found in object")
+	}
+	if len(ors) > 1 {
```

Comment:
wondering whether this is a problematic constraint. There can be other owners IMO.

Reply:
Indeed. I fixed this mainly in 7530710.